src/types/index.ts

```diff
@@ -592,6 +592,35 @@ export class StarknetContract {
         return this.adaptOutput(functionName, executed.stdout.toString());
     }
 
+    /**
+     * Computes L1-to-L2 message fee estimation
+     * @param {string} functionName Function name for entry point selector
+     * @param {StringMap} args - Arguments to Starknet contract function
+     * @returns Fee estimation
+     */
+    async estimateMessageFee(functionName: string, args: StringMap) {
+        // Check if functionName is annotated with @l1_handler
+        const func = <starknet.CairoFunction>this.abi[functionName];
+
+        if (!func?.type || func.type.toString() !== "l1_handler") {
+            throw new StarknetPluginError(
+                `Can not estimate message fee on "${functionName}" - not an @l1_handler`
+            );
+        }
+        const adaptedInput = this.adaptInput(functionName, args);
+        // Remove value of from_address from the input array
+        const keys = Object.keys(args);
+        const index = keys.indexOf("from_address");
+        if (index > -1) adaptedInput.splice(index, 1);
```

Comment:
I'd say the current approach is wrong because we are modifying the elements of `adaptedInput` based on their position in `args` (which is not an ordered structure - it's an object). It can be simplified and corrected as:
```suggestion
        const fromAddress = adaptedInput.shift();
```

Since shift pops the first element and reutrns it (we assume it to be fromAddress).
And later `args.fromAddress` needn't be used.

Reply:
https://github.com/Shard-Labs/starknet-hardhat-plugin/pull/312/commits/167c95b21c55677c3277fa6d26144c4ff099fe4a